samples/subsys/usb/hid-mouse/src/main.c

```diff
@@ -1,245 +1,117 @@
-#error "Unsupported board: sw0 devicetree alias is not defined"
-#endif
-
-#if !DT_NODE_EXISTS(LED0_NODE)
-#error "Unsupported board: led0 devicetree alias is not defined"
-#endif
-
-/*
- * Helper macro for initializing a gpio_dt_spec from the devicetree
- * with fallback values when the nodes are missing.
- */
-#define GPIO_SPEC(node_id) GPIO_DT_SPEC_GET_OR(node_id, gpios, {0})
-
-/*
- * Create gpio_dt_spec structures from the devicetree.
- */
-static const struct gpio_dt_spec sw0 = GPIO_SPEC(SW0_NODE),
-	sw1 = GPIO_SPEC(SW1_NODE),
-	sw2 = GPIO_SPEC(SW2_NODE),
-	sw3 = GPIO_SPEC(SW3_NODE),
-	led0 = GPIO_SPEC(LED0_NODE);
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(main, LOG_LEVEL_INF);
 
+static const struct gpio_dt_spec led0 = GPIO_DT_SPEC_GET(DT_ALIAS(led0), gpios);
 static const uint8_t hid_report_desc[] = HID_MOUSE_REPORT_DESC(2);
-
-static uint8_t def_val[4];
-static volatile uint8_t status[4];
-static K_SEM_DEFINE(sem, 0, 1);	/* starts off "not available" */
-static struct gpio_callback callback[4];
 static enum usb_dc_status_code usb_status;
 
-#define MOUSE_BTN_REPORT_POS	0
-#define MOUSE_X_REPORT_POS	1
-#define MOUSE_Y_REPORT_POS	2
-
 #define MOUSE_BTN_LEFT		BIT(0)
 #define MOUSE_BTN_RIGHT		BIT(1)
-#define MOUSE_BTN_MIDDLE	BIT(2)
+
+enum mouse_report_idx {
+	MOUSE_BTN_REPORT_IDX = 0,
+	MOUSE_X_REPORT_IDX = 1,
+	MOUSE_Y_REPORT_IDX = 2,
+	MOUSE_WHEEL_REPORT_IDX = 3,
+	MOUSE_REPORT_COUNT = 4,
+};
+
+static uint8_t report[MOUSE_REPORT_COUNT];
+static K_SEM_DEFINE(report_sem, 0, 1);
 
 static void status_cb(enum usb_dc_status_code status, const uint8_t *param)
 {
 	usb_status = status;
 }
 
-static void left_button(const struct device *gpio, struct gpio_callback *cb,
-			uint32_t pins)
+static void ALWAYS_INLINE rwup_if_suspended(void)
 {
-	int ret;
-	uint8_t state = status[MOUSE_BTN_REPORT_POS];
-
 	if (IS_ENABLED(CONFIG_USB_DEVICE_REMOTE_WAKEUP)) {
 		if (usb_status == USB_DC_SUSPEND) {
 			usb_wakeup_request();
 			return;
 		}
 	}
-
-	ret = gpio_pin_get(gpio, sw0.pin);
-	if (ret < 0) {
-		LOG_ERR("Failed to get the state of port %s pin %u, error: %d",
-			gpio->name, sw0.pin, ret);
-		return;
-	}
-
-	if (def_val[0] != (uint8_t)ret) {
-		state |= MOUSE_BTN_LEFT;
-	} else {
-		state &= ~MOUSE_BTN_LEFT;
-	}
-
-	if (status[MOUSE_BTN_REPORT_POS] != state) {
-		status[MOUSE_BTN_REPORT_POS] = state;
-		k_sem_give(&sem);
-	}
 }
 
-static void right_button(const struct device *gpio, struct gpio_callback *cb,
-			 uint32_t pins)
+static void input_cb(struct input_event *evt)
 {
-	int ret;
-	uint8_t state = status[MOUSE_BTN_REPORT_POS];
-
-	if (IS_ENABLED(CONFIG_USB_DEVICE_REMOTE_WAKEUP)) {
-		if (usb_status == USB_DC_SUSPEND) {
-			usb_wakeup_request();
-			return;
+	enum mouse_report_idx idx;
+	uint8_t value;
+
+	switch (evt->code) {
+	case INPUT_KEY_0:
+		rwup_if_suspended();
+		idx = MOUSE_BTN_REPORT_IDX;
+		value = report[idx];
+
+		if (evt->value) {
+			value |= MOUSE_BTN_LEFT;
+		} else {
+			value &= ~MOUSE_BTN_LEFT;
 		}
```

Comment:
you can use `WRITE_BIT(value, MOUSE_BTN_LEFT, evt->value);` here, makes it a bit more compact